File: src/main/java/wraith/fabricaeexnihilo/compatibility/rei/barrel/CompostCategory.java
package wraith.fabricaeexnihilo.compatibility.rei.barrel;

import me.shedaniel.math.Point;
import me.shedaniel.math.Rectangle;
import me.shedaniel.rei.api.client.gui.Renderer;
import me.shedaniel.rei.api.client.gui.widgets.Widget;
import me.shedaniel.rei.api.client.gui.widgets.Widgets;
import me.shedaniel.rei.api.client.registry.display.DisplayCategory;
import me.shedaniel.rei.api.common.category.CategoryIdentifier;
import me.shedaniel.rei.api.common.util.EntryStacks;
import net.minecraft.item.ItemStack;
import net.minecraft.text.Text;
import net.minecraft.util.Identifier;
import wraith.fabricaeexnihilo.compatibility.rei.GlyphWidget;
import wraith.fabricaeexnihilo.compatibility.rei.PluginEntry;

import java.util.ArrayList;
import java.util.List;

import static wraith.fabricaeexnihilo.FabricaeExNihilo.id;

public class CompostCategory implements DisplayCategory<CompostDisplay> {

    public static Identifier ARROW = id("textures/gui/rei/glyphs.png");
    public static int ARROW_U = 0;
    public static int ARROW_V = 16;
    private final ItemStack icon;
    private final String name;

    public CompostCategory(ItemStack icon, String name) {
        this.icon = icon;
        this.name = name;
    }

    @Override
    public CategoryIdentifier<? extends CompostDisplay> getCategoryIdentifier() {
        return PluginEntry.COMPOSTING;
    }

    @Override
    public int getDisplayHeight() {
        return 36;
    }

    @Override
    public Renderer getIcon() {
        return EntryStacks.of(icon);
    }

    @Override
    public Text getTitle() {
        return Text.translatable(this.name);
    }

    @Override
    public List<Widget> setupDisplay(CompostDisplay display, Rectangle bounds) {
        Point startPoint = new Point(bounds.getCenterX() - 41, bounds.getCenterY() - 13);
        var widgets = new ArrayList<Widget>();
        widgets.add(Widgets.createRecipeBase(bounds));

        widgets.add(new GlyphWidget(bounds, startPoint.x + 32, startPoint.y + 4, 16, 16, ARROW, ARROW_U, ARROW_V));

        widgets.add(Widgets.createSlot(new Point(startPoint.x + 61, startPoint.y + 5)).entries(display.input).markInput());
        widgets.add(Widgets.createSlot(new Point(startPoint.x + 4, startPoint.y + 5)).entries(display.result).markOutput());
        return widgets;
    }

}

File: src/main/kotlin/exnihilofabrico/compatibility/rei/barrel/CompostCategory.kt
package exnihilofabrico.compatibility.rei.barrel

import exnihilofabrico.ExNihiloFabrico
import exnihilofabrico.compatibility.rei.GlyphWidget
import exnihilofabrico.compatibility.rei.PluginEntry
import exnihilofabrico.id
import exnihilofabrico.util.asREIEntry
import exnihilofabrico.util.getExNihiloItemStack
import me.shedaniel.math.api.Rectangle
import me.shedaniel.rei.api.EntryStack
import me.shedaniel.rei.api.RecipeCategory
import me.shedaniel.rei.gui.widget.EntryWidget
import me.shedaniel.rei.gui.widget.RecipeBaseWidget
import me.shedaniel.rei.gui.widget.Widget
import java.util.function.Supplier

class CompostCategory: RecipeCategory<CompostDisplay> {

    override fun getIdentifier() = PluginEntry.COMPOSTING
    override fun getLogo() = getExNihiloItemStack("oak_barrel").asREIEntry()
    override fun getCategoryName() = "Barrel Composting"


    override fun getDisplayHeight() = HEIGHT
    override fun getDisplayWidth(display: CompostDisplay) =
        WIDTH

    override fun setupDisplay(displaySupplier: Supplier<CompostDisplay>, bounds: Rectangle): MutableList<Widget> {
        val display = displaySupplier.get()
        val widgets = mutableListOf<Widget>(RecipeBaseWidget(bounds))

        val arrowWidget = GlyphWidget(bounds, bounds.minX + ARROW_OFFSET_X, bounds.minY + ARROW_OFFSET_Y, ARROW_WIDTH, ARROW_HEIGHT, ARROW, ARROW_U, ARROW_V)
        widgets.add(arrowWidget)

        val inputs = display.inputEntries[0]
        val outputs = display.outputEntries

        // Output
        widgets.add(EntryWidget.create(bounds.minX + BLOCK_X, bounds.minY + BLOCK_Y).entries(outputs))

        // Input
        inputs.forEachIndexed { index, output ->
            widgets.add(
                EntryWidget.create(
                    bounds.minX + INPUT_X + (index % INPUT_SLOTS_X)*18,
                    bounds.minY + INPUT_Y + (index / INPUT_SLOTS_X)*18).entry(output))
        }
        // Fill in the empty spots
        for(index in inputs.size until MAX_INPUT) {
            widgets.add(
                EntryWidget.create(
                    bounds.minX + INPUT_X + (index % INPUT_SLOTS_X)*18,
                    bounds.minY + INPUT_Y + (index / INPUT_SLOTS_X)*18).entry(EntryStack.empty()))
        }

        return widgets
    }

    companion object {
        val ARROW = id("textures/gui/rei/glyphs.png")

        val INPUT_SLOTS_X = maxOf(ExNihiloFabrico.config.modules.REI.compostNumCols, 1)
        val INPUT_SLOTS_Y = maxOf(ExNihiloFabrico.config.modules.REI.compostNumRows, 1)

        val MARGIN = 6

        val WIDTH = 2*18 + INPUT_SLOTS_X*18 + MARGIN*2
        val HEIGHT = INPUT_SLOTS_Y*18 + MARGIN*2

        val BLOCK_X = MARGIN
        val ARROW_OFFSET_X = BLOCK_X + 18
        val INPUT_X = ARROW_OFFSET_X + 18


        val BLOCK_Y = MARGIN + (HEIGHT - 2*MARGIN)/2 - 9
        val ARROW_OFFSET_Y = BLOCK_Y
        val INPUT_Y = MARGIN

        val ARROW_WIDTH = 16
        val ARROW_HEIGHT= 16
        val ARROW_U = 0
        val ARROW_V= 16

        val MAX_INPUT = INPUT_SLOTS_X * INPUT_SLOTS_Y
    }

}